Public/Src/Engine/Scheduler/TempCleaner.cs

```diff
@@ -126,17 +126,17 @@ public void Dispose()
                 m_disposed = true;
 
                 m_cancellationSource.Cancel();
-                bool terminated = m_cleanerThread.Join(TimeSpan.FromMinutes(1));
+                m_pathsToDelete.CompleteAdding();
+                m_cleanerThread.Join();

```

Comment:
>rThread.Join(); [](start = 24, length = 15)

I thought the design was that we should not hold up the build any longer than a minute to clean temp directories after the build completes. Cleaning will be picked up by the next build in the background or when immediately needed.

Reply:
You're right, this seems to be a change in the design. I'm still wanting to make the build ends quickly without waiting for the temp cleaner to finish. That's the reason I put the timeout in `Thread.Join`. However, that doesn't play nicely with the blocking collection. 

Currently, our hope for quick termination relies on the cancellation of directory/file deletion. This should be OK for our selfhost. But apparently, that is not sufficient for lab/cloud build. I think we need to understand better why cancellation is not so great there.